File: python/00117_populating_next_right_pointers_in_each_node_II.py
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return None

        from collections import deque
        queue = deque([root])

        while queue:
            size = len(queue)
            for i in range(size):
                current = queue.popleft()
                # Connect to the next node if it’s not the last in the level
                if i < size - 1:
                    current.next = queue[0]
                else:
                    current.next = None
                # Add children
                if current.left:
                    queue.append(current.left)
                if current.right:
                    queue.append(current.right)

        return root

File: java/00117_populating_next_right_pointers_in_each_node_II.java
class Solution {
    public Node connect(Node root) {

        if (root == null) {
            return root;
        }

        Node pointerNode = new Node (-1);
        pointerNode.next = root;

        Queue <Node> queue = new LinkedList <> ();

        queue.add (root);

        while (!queue.isEmpty()) {
            int size = queue.size ();

            int j=0;
            while (j<size) {
                Node current = queue.poll ();
                j++;
                Node nextNode = null;
                if (j==size) {
                    current.next = nextNode;
                } else {
                    nextNode = queue.peek();
                    current.next = nextNode;
                }
                if (current.left!=null) {
                    queue.add (current.left);
                }
                if (current.right!=null) {
                    queue.add (current.right);
                }
            }
        }

        return pointerNode.next;
    }
}